synapse/handlers/oidc_handler.py

```diff
@@ -904,19 +907,21 @@ async def _map_userinfo_to_user(
 
         user_id = UserID(localpart, self._hostname)
         if await self._datastore.get_users_by_id_case_insensitive(user_id.to_string()):
-            # This mxid is taken
-            raise MappingException(
-                "mxid '{}' is already taken".format(user_id.to_string())
+            if self._merge_with_existing_users:
+                registered_user_id = user_id.to_string()
```

Comment:
it is not a given that the registered user id is the same as the id we looked up (note: "_case_insensitive".) `get_users_by_id_case_insensitive` returns a dict of potiental matches. please use it.